File: survival-common/src/main/java/net/survival/world/chunk/VirtualAllocationUnit.java
package net.survival.world.chunk;

import java.nio.ByteBuffer;

class VirtualAllocationUnit
{
    public static final long MAX_LENGTH = 65536;
    public static final int STRUCTURE_SIZE = 17;

    public long address;
    public long length;
    public boolean allocated;

    public VirtualAllocationUnit() {}

    public VirtualAllocationUnit(long address, long length, boolean allocated) {
        this.address = address;
        this.length = length;
        this.allocated = allocated;
    }

    public void readFrom(ByteBuffer buffer) {
        address = buffer.getLong();
        length = buffer.getLong();
        allocated = (buffer.get() == 1);
    }

    public void writeTo(ByteBuffer buffer) {
        buffer.putLong(address);
        buffer.putLong(length);
        buffer.put(allocated ? (byte) 1 : (byte) 0);
    }
}

File: survival-common/src/main/java/net/survival/block/io/VirtualAllocationUnit.kt
package net.survival.block.io

import java.nio.ByteBuffer

internal class VirtualAllocationUnit {
    companion object {
        const val MAX_LENGTH: Long = 16777216
        const val STRUCTURE_SIZE = 17
    }

    var address: Long = 0
    var length: Long = 0
    var allocated = false

    constructor()

    constructor(address: Long, length: Long, allocated: Boolean) {
        this.address = address
        this.length = length
        this.allocated = allocated
    }

    fun readFrom(buffer: ByteBuffer) {
        address = buffer.long
        length = buffer.long
        allocated = buffer.get().toInt() == 1
    }

    fun writeTo(buffer: ByteBuffer) {
        buffer.putLong(address)
        buffer.putLong(length)
        buffer.put(if (allocated) 1.toByte() else 0.toByte())
    }
}